File: C/circ_buf.c
/****************************
 *    Copyright (c) 2010    *
 *    Keith Cullen          *
 ****************************/

/*
 *  A queue implemented using a contiguous buffer with separate indices for reading and writing.
 *
 *  Bytes are copied in to and out from the buffer.
 *
 *  The head index is the next location to be written. It is incremented when items
 *  are added and wraps to zero when it reaches the end of the linear buffer.
 *
 *  The tail index is the next location to be read. It is incremented when items
 *  are removed and wraps to zero when it reaches the end of the linear buffer.
 *
 *  When the head index is equal to the tail index, the circular buffer is empty.
 *  When the head index is one less than the tail index, the circular buffer is full.
 */

#include <string.h>
#include "circ_buf.h"

void circ_buf_init(circ_buf_t *cb, char *buf, unsigned len)
{
    memset(buf, 0, len);
    cb->head = 0;
    cb->tail = 0;
    cb->buf = buf;
    cb->len = len;
}

/*  read data but don't update tail
 *  (2 consecutive peek operations with the same arguments will produce the same result)
 *  returns number of bytes read
 */
int circ_buf_peek(circ_buf_t *cb, char *buf, unsigned len)
{
    unsigned tail = cb->tail;
    unsigned num = 0;
    int ret = 0;

    while (1)
    {
        num = circ_buf_count_to_end_(cb, tail);
        if (len < num)
            num = len;
        if (num <= 0)
            break;
        memcpy(buf, cb->buf + tail, num);
        tail = circ_buf_wrap_index(cb, tail + num);
        buf += num;
        len -= num;
        ret += num;
    }
    return ret;
}

/*  returns number of bytes read
 */
int circ_buf_consume(circ_buf_t *cb, unsigned len)
{
    unsigned num = 0;
    int ret = 0;

    while (1)
    {
        num = circ_buf_count_to_end(cb);
        if (len < num)
            num = len;
        if (num <= 0)
            break;
        cb->tail = circ_buf_wrap_index(cb, cb->tail + num);
        len -= num;
        ret += num;
    }
    return ret;
}

/*  returns number of bytes read
 */
int circ_buf_read(circ_buf_t *cb, char *buf, unsigned len)
{
    unsigned num = 0;
    int ret = 0;

    while (1)
    {
        num = circ_buf_count_to_end(cb);
        if (len < num)
            num = len;
        if (num <= 0)
            break;
        memcpy(buf, cb->buf + cb->tail, num);
        cb->tail = circ_buf_wrap_index(cb, cb->tail + num);
        buf += num;
        len -= num;
        ret += num;
    }
    return ret;
}

/*  returns number of bytes written
 */
int circ_buf_write(circ_buf_t *cb, const char *buf, unsigned len)
{
    unsigned num = 0;
    int ret = 0;

    while (1)
    {
        num = circ_buf_space_to_end(cb);
        if (len < num)
            num = len;
        if (num <= 0)
            break;
        memcpy(cb->buf + cb->head, buf, num);
        cb->head = circ_buf_wrap_index(cb, cb->head + num);
        buf += num;
        len -= num;
        ret += num;
    }
    return ret;
}

File: C#/CircBuf.cs
using System;

namespace Circular
{
    class CircBuf<T>
    {
        public int Head { get; set; }
        public int Tail { get; set; }
        public int Len  { get; set; }
        public T[] Buf  { get; set; }

        private bool powerOf2(int x)
        {
            for (int i = 1; i < 12; i++)
            {
                if (x == (1 << i))
                {
                    return true;
                }
            }
            return false;
        }

        public CircBuf(int len)
        {
            if (!powerOf2(len))
            {
                throw new ArgumentOutOfRangeException("len", "len must be an integer power of 2.");
            }
            Head = 0;
            Tail = 0;
            Len = len;
            Buf = new T[len];
        }

        private int countToEnd(int tail)
        {
            int countEndLinearBuf = Len - tail;
            int countEndCircBuf = (Head + countEndLinearBuf) & (Len - 1);
            return countEndCircBuf < countEndLinearBuf ? countEndCircBuf : countEndLinearBuf;
        }

        private int countToEnd()
        {
            return countToEnd(Tail);
        }

        private int spaceToEnd()
        {
            int spaceEndLinearBuf = Len - Head;
            int spaceEndCircBuf = (Tail + spaceEndLinearBuf - 1) & (Len - 1);
            return spaceEndLinearBuf < spaceEndCircBuf ? spaceEndLinearBuf : spaceEndCircBuf;
        }

        public int Count()
        {
            return (Head - Tail) & (Len - 1);
        }

        public int Space()
        {
            return (Tail - Head - 1) & (Len - 1);
        }

        public int Pop(ref T t)
        {
            if (Count() == 0)
            {
                return 0;
            }
            t = Buf[Tail];
            Tail = (Tail + 1) & (Len - 1);
            return 1;
        }

        public int Push(T t)
        {
            if (Space() == 0)
            {
                return 0;
            }
            Buf[Head] = t;
            Head = (Head + 1) & (Len - 1);
            return 1;
        }

        public int Read(T[] t)
        {
            if (t.Length <= 0)
            {
                throw new ArgumentOutOfRangeException("t", "Length must be greater than zero.");
            }

            int len = t.Length;
            int ret = 0;
            int j = 0;
            while (true)
            {
                int num = countToEnd();
                if (len < num)
                {
                    num = len;
                }
                if (num <= 0)
                {
                    break;
                }
                for (int i = 0; i < num; i++)
                {
                    t[i + j] = Buf[i + Tail];
                }
                Tail = (Tail + num) & (Len - 1);
                j += num;
                len -= num;
                ret += num;
            }
            return ret;
        }

        public int Write(T[] t)
        {
            if (t.Length <= 0)
            {
                throw new ArgumentOutOfRangeException("t", "Length must be greater than zero.");
            }

            int len = t.Length;
            int ret = 0;
            int j = 0;
            while (true)
            {
                int num = spaceToEnd();
                if (len < num)
                {
                    num = len;
                }
                if (num <= 0)
                {
                    break;
                }
                for (int i = 0; i < num; i++)
                {
                    Buf[i + Head] = t[i + j];
                }
                Head = (Head + num) & (Len - 1);
                j += num;
                len -= num;
                ret += num;
            }
            return ret;
        }

        public int Peek(T[] t)
        {
            if (t.Length <= 0)
            {
                throw new ArgumentOutOfRangeException("t", "Length must be greater than zero.");
            }

            int len = t.Length;
            int tail = Tail;
            int ret = 0;
            int j = 0;
            while (true)
            {
                int num = countToEnd(tail);
                if (len < num)
                {
                    num = len;
                }
                if (num <= 0)
                {
                    break;
                }
                for (int i = 0; i < num; i++)
                {
                    t[i + j] = Buf[i + tail];
                }
                tail = (tail + num) & (Len - 1);
                j += num;
                len -= num;
                ret += num;
            }
            return ret;
        }

        public int Consume(T[] t)
        {
            if (t.Length <= 0)
            {
                throw new ArgumentOutOfRangeException("t", "Length must be greater than zero.");
            }

            int len = t.Length;
            int ret = 0;
            while (true)
            {
                int num = countToEnd();
                if (len < num)
                {
                    num = len;
                }
                if (num <= 0)
                {
                    break;
                }
                Tail = (Tail + num) & (Len - 1);
                len -= num;
                ret += num;
            }
            return ret;
        }
    }
}
